File: packages/core/src/planner.ts
import type { AgentModelPreset, AgentTool, PlannedTask, TaskDomain } from "./types";

export const defaultModels: Record<AgentTool, string> = {
  "claude-code": "claude-sonnet-4",
  codex: "gpt-5.2",
  opencode: "anthropic/claude-sonnet-4",
};

export const defaultPreset: AgentModelPreset = {
  name: "Default Split",
  mapping: {
    frontend: { agent: "opencode", model: defaultModels.opencode },
    backend: { agent: "opencode", model: defaultModels.opencode },
    shared: { agent: "opencode", model: defaultModels.opencode },
    qa: { agent: "opencode", model: defaultModels.opencode },
  },
};

export function validatePreset(preset: AgentModelPreset): AgentModelPreset {
  const domains: TaskDomain[] = ["frontend", "backend", "shared", "qa"];
  for (const domain of domains) {
    const selection = preset.mapping[domain];
    if (!selection?.agent || !selection.model) {
      throw new Error(`Preset ${preset.name} must define agent and model for ${domain}`);
    }
    if (!["claude-code", "codex", "opencode"].includes(selection.agent)) {
      throw new Error(`Unsupported agent for ${domain}: ${selection.agent}`);
    }
  }
  return preset;
}

export function planTasks(spec: import("./types").SpecDocument, preset: AgentModelPreset = defaultPreset): PlannedTask[] {
  validatePreset(preset);
  const actionableSections = spec.sections.filter((section) => section.depth <= 3 && section.content.length > 0);
  const counters: Record<TaskDomain, number> = { frontend: 0, backend: 0, shared: 0, qa: 0 };
  const tasks = actionableSections.map((section) => {
    const domain = inferDomain(section.slug);
    counters[domain] += 1;
    const id = `${domain}-${String(counters[domain]).padStart(2, "0")}`;
    const selection = preset.mapping[domain];
    return {
      id,
      domain,
      title: section.title,
      description: section.content,
      dependsOn: [] as string[],
      agent: selection.agent,
      model: selection.model,
      status: "pending" as const,
      branchName: `feature/${id}`,
    };
  });

  const apiTask = tasks.find((task) => task.title.toLowerCase().includes("api contract"));
  if (apiTask) {
    for (const task of tasks) {
      if (task.domain === "frontend" && task.id !== apiTask.id) task.dependsOn.push(apiTask.id);
    }
  }
  return tasks;
}

export function inferDomain(slug: string): TaskDomain {
  if (slug.includes("frontend") || slug.includes("ui") || slug.includes("tui")) return "frontend";
  if (slug.includes("backend") || slug.includes("api") || slug.includes("database") || slug.includes("server")) return "backend";
  if (slug.includes("test") || slug.includes("qa") || slug.includes("verification")) return "qa";
  return "shared";
}

export function formatTaskList(tasks: PlannedTask[]) {
  return tasks.map((task) => {
    const deps = task.dependsOn.length > 0 ? ` depends on ${task.dependsOn.join(", ")}` : "";
    return `${task.id} [${task.domain}] ${task.title} -> ${task.agent}:${task.model}${deps}`;
  }).join("\n");
}

File: internal/planner/planner.go
package planner

import (
	"bufio"
	"crypto/sha1"
	"encoding/hex"
	"os"
	"strings"

	"github.com/forkestra-cli/forkestra/internal/model"
)

func Load(path string) ([]model.PlannedTask, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	var tasks []model.PlannedTask
	scanner := bufio.NewScanner(f)
	var title, body string
	flush := func() {
		if title == "" || strings.TrimSpace(body) == "" {
			return
		}
		domain := domainFor(title)
		id := string(domain) + "-" + hash(title)[:8]
		tasks = append(tasks, model.PlannedTask{ID: id, Title: title, Description: strings.TrimSpace(body), Domain: domain, Agent: model.OpenCode, Model: "anthropic/claude-sonnet-4", Status: model.Pending, BranchName: "feature/" + id})
		title = ""
		body = ""
	}
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "#") {
			flush()
			title = strings.TrimSpace(strings.TrimLeft(line, "#"))
		} else if title != "" {
			body += line + "\n"
		}
	}
	flush()
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	for _, t := range tasks {
		if strings.Contains(strings.ToLower(t.Title), "api contract") {
			for i := range tasks {
				if tasks[i].Domain == model.Frontend && tasks[i].ID != t.ID {
					tasks[i].DependsOn = append(tasks[i].DependsOn, t.ID)
				}
			}
		}
	}
	return tasks, nil
}
func domainFor(title string) model.Domain {
	s := strings.ToLower(title)
	switch {
	case strings.Contains(s, "frontend") || strings.Contains(s, "ui") || strings.Contains(s, "tui"):
		return model.Frontend
	case strings.Contains(s, "backend") || strings.Contains(s, "api") || strings.Contains(s, "database") || strings.Contains(s, "server"):
		return model.Backend
	case strings.Contains(s, "test") || strings.Contains(s, "qa") || strings.Contains(s, "verification"):
		return model.QA
	default:
		return model.Shared
	}
}
func hash(s string) string { h := sha1.Sum([]byte(s)); return hex.EncodeToString(h[:]) }
func Format(tasks []model.PlannedTask) string {
	var lines []string
	for _, t := range tasks {
		lines = append(lines, t.ID+" ["+string(t.Domain)+"] "+t.Title+" -> "+string(t.Agent)+":"+t.Model)
	}
	return strings.Join(lines, "\n")
}
